java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java

```diff
@@ -0,0 +1,160 @@
+import org.netbeans.api.db.explorer.JDBCDriverManager;
+import org.netbeans.modules.java.lsp.server.protocol.NbCodeLanguageClient;
+import org.netbeans.modules.java.lsp.server.protocol.QuickPickItem;
+import org.netbeans.modules.java.lsp.server.protocol.ShowInputBoxParams;
+import org.netbeans.modules.java.lsp.server.protocol.ShowQuickPickParams;
+import org.openide.filesystems.FileObject;
+import org.openide.filesystems.URLMapper;
+import org.openide.util.NbBundle;
+
+/**
+ *
+ * @author Jan Horvath
+ */
+@NbBundle.Messages({
+    "MSG_EnterDbUrl=Enter DB URL",
+    "MSG_EnterUsername=Enter Username",
+    "MSG_EnterPassword=Enter Password",
+    "MSG_SelectDriver=Select db driver",
+    "MSG_DriverNotFound=Driver not found",
+    "MSG_ConnectionAdded=Connection added",
+    "MSG_ConnectionFailed=Connection failed",
+    "MSG_SelectSchema=Select Database Schema"
+})
+public class DatabaseCommands {
+
+    public static CompletableFuture<Object> addDbConnection(NbCodeLanguageClient client) {
+        JDBCDriver[] drivers = JDBCDriverManager.getDefault().getDrivers();
+        List<QuickPickItem> items = new ArrayList<>();
+        for (int i = 0; i < drivers.length; i++) {
+            URL[] jars = drivers[i].getURLs();
+            if (jars != null && jars.length > 0) {
+                FileObject jarFO = URLMapper.findFileObject(jars[0]);
+                if (jarFO != null && jarFO.isValid()) {
+                    items.add(
+                            new QuickPickItem(drivers[i].getName(), null, drivers[i].getDisplayName() + " (" + drivers[i].getClassName() + ")", false, i) // NOI18N
+                    );
+                }
+            }
+        }
+
+        return client.showQuickPick(new ShowQuickPickParams(Bundle.MSG_SelectDriver(), false, items))
+                .thenApply(selectedItems -> {
+                    if (selectedItems == null) {
+                        return null;
+                    }
+                    if (!selectedItems.isEmpty()) {
+                        int i = ((Double) selectedItems.get(0).getUserData()).intValue();
+                        JDBCDriver driver = drivers[i];
+                        client.showInputBox(new ShowInputBoxParams(
+                                Bundle.MSG_EnterDbUrl(), "")).thenAccept((u) -> {
+                            if (u == null) {
+                                return;
+                            }
+                            client.showInputBox(new ShowInputBoxParams(
+                                    Bundle.MSG_EnterUsername(), "")).thenAccept((username) -> { //NOI18N
+                                if (username == null) {
+                                    return;
+                                }
+                                client.showInputBox(new ShowInputBoxParams(
+                                        Bundle.MSG_EnterPassword(), "", true)).thenAccept((password) -> { //NOI18N
+                                    if (password == null) {
```

Comment:
Is this intentional, that you open 3 consecutive prompts for the connection parameters? I think the current approach on the Swing side useing a single dialog is more user friendly (and the test connection button is a bonus :-))

Reply:
Note: this is an LSP-specific command implementation -- NetBeans have its standard Swing dialogs implemented in the DB module. Sadly LSP protocol does not allow (yet ?) for rich dialogs, so we need to approximate, I guess. 

Some DB - related commands (i.e. connect, disconnect) were bridged directly to NB implementations although that requried some fixes in NB codebase. So LSP module contains minimal stuff needed - as much as it can be done is done directly in IDE codebase.

Reply:
Yeah, I assumed that much - thanks for the confirmation.